src/ar_http_iface_cowboy_handler.erl

```diff
@@ -659,21 +665,139 @@ do_handle(Method, [<<"height">>], Req)
+		{custom, Domain} -> handle_serve_tx_apex_1(ApexDomain, TxHash, Req)
+	end.
+
+get_domain_labeling(ApexDomain, Domain) ->
+	case binary:match(Domain, ApexDomain) of
+		{0, _} ->
+			apex;
+		{N, _} ->
+			Label = binary:part(Domain, {0, N-1}),
+			{labeled, Label};
+		nomatch ->
+			get_domain_labeling_1(Domain)
+	end.
+
+get_domain_labeling_1(Domain) ->
+	CustomDomains = ar_meta_db:get(custom_domains),
+	case lists:member(Domain, CustomDomains) of
+		true -> {custom, Domain};
+		false -> error('FIXME')
+	end.
+
+redirect_to_apex(ApexDomain, Req) ->
+	Scheme = cowboy_req:scheme(Req),
+	Port = integer_to_binary(cowboy_req:port(Req)),
+
+	Location =
+		<<Scheme/binary, "://",
+			ApexDomain/binary, ":",
+			Port/binary, "/">>,
+
+	{301, #{<<"location">> => Location}, <<>>, Req}.
+
+resolve_custom_domain_tx(Domain, Req) ->
+	case inet_res:lookup("_arweave." ++ binary_to_list(Domain), in, txt) of
+		[] ->
+			{502, #{}, <<"Domain improperly configured">>, Req};
+		[RecordChunks|_] ->
+			TxHash = list_to_binary(lists:concat(RecordChunks)),
+			handle(<<"GET">>, [<<"tx">>, TxHash, <<"data.html">>], Req)
+	end.
+
+handle_serve_tx_apex_1(ApexDomain, TxHash, Req) ->
+	BlockHash = ?OK(get_tx_block_hash(TxHash)),
+	redirect_to_labeled_tx(ApexDomain, BlockHash, TxHash, Req).
+
+handle_serve_tx_labeled_1(ApexDomain, Label, TxHash, MaybeExt, Req) ->
+	BlockHash = ?OK(get_tx_block_hash(TxHash)),
+	handle_serve_tx_labeled_2(ApexDomain, Label, BlockHash, TxHash, MaybeExt, Req).
+
+get_tx_block_hash(TxHash) ->
+	case ar_tx_search:get_tags_by_id(
+		whereis(http_search_node),
+		ar_util:decode(TxHash),
+		infinity
+	) of
+		{ok, PseudoTags} ->
+			case proplists:get_value(<<"block_indep_hash">>, PseudoTags) of
+				undefined -> not_found;
+				BlockHash -> {ok, BlockHash}
+			end;
+		{error, _} ->
+			not_found
+	end.
+
+handle_serve_tx_labeled_2(ApexDomain, Label, BlockHash, TxHash, MaybeExt, Req) ->
+	ExpectedLabel = derive_tx_label(TxHash, BlockHash),
+	case Label of
+		ExpectedLabel -> handle_serve_tx_labeled_3(TxHash, MaybeExt, Req);
+		_ -> redirect_to_labeled_tx(ApexDomain, BlockHash, TxHash, Req)
+	end.
+
+derive_tx_label(TxHash, BlockHash) ->
+	State0 = crypto:hash_init(sha256),
+	State1 = crypto:hash_update(State0, ar_util:decode(TxHash)),
+	State2 = crypto:hash_update(State1, ar_util:decode(BlockHash)),
+	Digest = crypto:hash_final(State2),
+	binary:part(ar_base32:encode(Digest), {0, 12}).
+
+handle_serve_tx_labeled_3(TxHash, <<>>, Req) ->
+	handle_serve_tx_labeled_3(TxHash, <<".html">>, Req);
+handle_serve_tx_labeled_3(TxHash, <<".", Ext/binary>>, Req) ->
+	handle(<<"GET">>, [<<"tx">>, TxHash, <<"data.", Ext/binary>>], Req);
+handle_serve_tx_labeled_3(_, _, Req) ->
 	not_found(Req).
 
+redirect_to_labeled_tx(ApexDomain, BlockHash, TxHash, Req) ->
+	Scheme = cowboy_req:scheme(Req),
+	Label = derive_tx_label(TxHash, BlockHash),
+	Port = integer_to_binary(cowboy_req:port(Req)),
+
+	Location =
```

Comment:
Even though we don't do Erlang standard indentation, I think we can do better than this. How about this:
```
	Location = <<
		Scheme/binary, "://",
		Label/binary, ".",
		ApexDomain/binary, ":",
		Port/binary, "/",
		TxHash/binary
	>>,
```
Or this:
```
	Location = <<
			Scheme/binary, "://",
			Label/binary, ".",
			ApexDomain/binary, ":",
			Port/binary, "/",
			TxHash/binary
		>>,
```